[TCL30-9, TCL30-5]  Fetch data from Wikipedia 'nearby' API and render it on the map

katia-utochkina opened the pull request:
## Description
This PR adds ``Spinner``, ``ErrorMessage``, ``WikiResults`` components to the app. Along with that, it fetches Paris locations with Wikipedia entries from the ‘nearby’ API proxy endpoint and makes the data returned available to the ``<Map/>`` component. This PR also adds custom pins to the map, using data fetched from "nearby" API

## Related Issue

Closes [TCL30-5](https://the-collab-lab.atlassian.net/browse/TCL30-5) and [TCL30-9](https://the-collab-lab.atlassian.net/browse/TCL30-9)


## Acceptance Criteria

- When the app starts, a call should be made to the proxy endpoint using the latitude and longitude at the centre of the map

- The returned data should be saved to state or in some other way made available to rendering components 

- There are locations (e.g., the middle of the ocean) where we will not get any locations back from the API. The app should handle that case without generating errors

- Data fetched from "nearby" API should be used to render pins on the map

- Pins should be 24px x 24px in size and should have a little down caret on the bottom edge of the pin pointing to the location of the place

- Pins should show a scaled-down thumbnail image in cases when a thumbnail image is present

- Pins should render a default pin image when no thumbnail is returned

- Pins should flow in onto the map

## Type of Changes

<!-- Put an `✓` for the applicable box: -->

|     | Type                       |
| --- | -------------------------- |
|     | :bug: Bug fix              |
| ✓   | :sparkles: New feature     |
|     | :hammer: Refactoring       |
|     | :100: Add tests            |
|     | :link: Update dependencies |
|     | :scroll: Docs              |

## Updates

### Before

There was no access to the Wikipedia 'nearby' entries in the app
![before_upd](https://user-images.githubusercontent.com/16920088/126311118-b0ea9465-b3ff-424f-b15c-a7bc0f97532f.png)

### After

App components have access to the data fetched from the Wikipedia 'nearby' API. The returned data is used to render custom pins on the map
![after555](https://user-images.githubusercontent.com/16920088/126359298-940b8d21-d515-414b-965f-0e4f4b128e60.png)



## Testing Steps / QA Criteria

- From your terminal, pull down this branch - ``ku-mm-fetch-wiki-data``
-  Run ``npm start`` to launch the app
-  Inspect state for the expected data
![devtools](https://user-images.githubusercontent.com/16920088/126078476-e321e5cf-9bd3-42e3-b9f6-570ef43fa335.png)
-  Launch the app with a location that has been verified not to return any Wikipedia entries. For example,  replace ``const { lat, lng } = { lat: 48.8566, lng: 2.3522 }``  in the WikiResults component with `` const { lat, lng } = { lat: 8.7832, lng: 124.5085 }``
-  Make sure that for the locations verified to return Wikipedia entries, pins are rendered on the map
-  Make sure those pins show a scaled-down thumbnail image in cases when a thumbnail image is present and their size is 24x24
-  Make sure the pins render a default pin image when no thumbnail is returned
-  Make sure the pins flow in onto the map

katia-utochkina committed:
[TCL30-35]  Fetch data from Wikipedia 'nearby' API

-Add Spinner, ErrorMessage, WikiResults components
-Fetch locations
-Handle the case for locations not found
-Make fetched data available to Map component

katia-utochkina committed:
Update file

katia-utochkina committed:
Update file

katia-utochkina committed:
Remove redundant code

katia-utochkina committed:
Update file

alxmcr commented on src/App.css:
```diff
@@ -145,3 +145,26 @@ nav {
   text-decoration: underline;
 }
+
+@keyframes donut-spin {
```
I loved your spinner 💛! Congrats on your great job!

alxmcr commented on src/components/ErrorMessage.js:
```diff
@@ -0,0 +1,11 @@
+import React from 'react';
+
+const ErrorMessage = ({ error }) => (
```
Interesting idea! It could be a good way to display some errors.

alxmcr commented on src/components/Spinner.js:
```diff
@@ -0,0 +1,9 @@
+import React from 'react';
+
+const Spinner = () => (
```
I loved it 💛!

alxmcr commented on src/components/WikiResults.js:
```diff
@@ -0,0 +1,57 @@
+
+  const { lat, lng } = { lat: 48.8566, lng: 2.3522 };
+
+  useEffect(() => {
```
**_Only suggestions (There is no pressure 😄):_**

-  _What do you think about creating a custom hook in this case?_ 
- _Did you consider about pass `lat` and `lng` through some `context`?_

laurenmbeatty replied:
I like @alxmcr 's suggestion...It would be good to have this be at a higher level so you can pass the data down to both the Map and the List components.

alxmcr reviewed (approved):
When I was reviewing your great work:

- I was impressed with your `Spinner` and `ErrorMessage` component.
- Today I learned a lot: `status` state (loading, success, and error)

Great work, congrats @meganmckissack and @katia-utochkina 😄🎉!  

I have some suggestions, but they're not blocking😄. Approved ✔️!

alxmcr commented on src/components/WikiResults.js:
```diff
@@ -0,0 +1,57 @@
+
+const WikiResults = () => {
+  const [data, setData] = useState([]);
+  const [status, setStatus] = useState('idle');
```
I'm curious, what does the 'idle' word mean 😄? and great your options for status (loading, success, and error)

meganmckissack replied:
@alxmcr I can double check with Katia but it's a `loading` state so we could change the wording to be more clear.  Thank you!

katia-utochkina replied:
hey @alxmcr, we have an element of state to track the status of our search.  ``idle`` here means we haven't started doing the search yet.😄

alxmcr replied:
Thanks, I'm a little confused but today I learned: `idle` means 'not yet'`.

- _(A little suggestion - no pressure_ 😄) What do you think about `''` (empty string) or 'pending'?

katia-utochkina replied:
hey @alxmcr 😄  not sure ``pending`` would work here. ``pending`` is synonymous with  ``loading``  which is another element of state, used in the code, to indicate we're in the process of loading search results

katia-utochkina committed:
[TCL30-9] Render pins on the map from "nearby" API

- Add Pin component
- Add default pin image to Pin component
- Render pins on the map using fetched data
- Center Spinner

katia-utochkina committed:
Fix spinner styles

katia-utochkina committed:
Update file

katia-utochkina committed:
Move a div wrapping spinner to Spinner component

katia-utochkina committed:
Add a wrapping div to Spinner component

katia-utochkina committed:
Move a div wrapping spinner to Spinner component

katia-utochkina committed:
Add a wrapping div to Spinner component

katia-utochkina committed:
Give spinner container a class

katia-utochkina committed:
Fix styling for images

katia-utochkina committed:
Animate pins

katia-utochkina committed:
Add placeholder image

katia-utochkina committed:
Remove marker const

meganmckissack committed:
updated size properties for the icon images and removed width and height properties in the css to fix the map controls size bug

katia-utochkina committed:
Remove error from dependecy array

katia-utochkina committed:
Remove error from dependency array

katia-utochkina committed:
Update file

katia-utochkina committed:
Merge branch 'ku-mm-render-map-pins' into ku-mm-fetch-wiki-data

katia-utochkina committed:
Add Pin component

katia-utochkina committed:
Update pin styles

katia-utochkina committed:
Render pins after map is loaded

katia-utochkina committed:
Add reduced motion media query

jssckbl commented:
I am in the "Testing Steps" and am coming up with an error when I perform the following step:

`Launch the app with a location that has been verified not to return any Wikipedia entries. For example, replace const { lat, lng } = { lat: 48.8566, lng: 2.3522 } in the WikiResults component with const { lat, lng } = { lat: 8.7832, lng: 124.5085 }`

<img width="465" alt="testingError" src="https://user-images.githubusercontent.com/38255956/126537275-5a8cd36d-b0e8-4b73-85fb-8a63b0102044.png">

Any suggestions?

laurenmbeatty commented on src/components/Map.js:
```diff
@@ -18,10 +20,32 @@ const Map = () => {
+            yesIWantToUseGoogleMapApiInternals={true}
+            onGoogleApiLoaded={({ map, maps }) => handleApiLoaded(map, maps)}
+          >
+            {data && data.query.pages.length > 0 && loaded
```
Good eye, @jssckbl . That error is probably happening because `query` is undefined above ^

laurenmbeatty replied:
You all could use [optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) to solve for that.
 `data?.query?.pages?.length`

katia-utochkina replied:
Thank you, @jssckbl. It's fixed now😄

katia-utochkina committed:
Add optional chaining operator to check if query obj and pages array exist in the data obj

meganmckissack committed:
Add Context API

meganmckissack committed:
remove unused code

meganmckissack committed:
Add custom  hook

katia-utochkina committed:
Remove unused file

katia-utochkina committed:
Remove console.log

katia-utochkina committed:
Add alt attribute to all marker images

katia-utochkina committed:
Add title, id, aria-label to svg icon to make it accessible

katia-utochkina committed:
Rename title prop to alt prop

katia-utochkina committed:
Move SearchResults to context folder

laurenmbeatty commented on src/context/SearchResults.js:
```diff
@@ -0,0 +1,35 @@
+export const SearchResultsContext = createContext();
+
+const SearchResults = ({ children }) => {
+  const { lat, lng } = { lat: 48.8566, lng: 2.3522 };
```
I wonder if there is a way to set the `lat` and `lng` in only one place. Right now it's being done here, and in the `useFetch`. Maybe `useFetch` should just take in the lat and lng.

katia-utochkina replied:
Hi @laurenmbeatty. There's been so much refactoring going on here, we missed that. Yes, having it one place only is a good idea.  Fixed now😄

laurenmbeatty commented on src/hooks/useFetch.js:
```diff
@@ -0,0 +1,39 @@
+  }, [lat, lng, url]);
+
+  return { data, status, error };
+};
```
i like this custom hook 👍

laurenmbeatty reviewed (commented):
Looks like you've handled edge cases really well! Just a question from me about setting the lat/lng in 2 different places.

katia-utochkina committed:
Give useFetch lat and lng parameters

katia-utochkina committed:
Call useFetch hook with lat, lng arguments

segdeha commented on src/App.css:
```diff
@@ -47,6 +47,21 @@ textarea {
   font: inherit;
 }
 
+@media (prefers-reduced-motion: reduce) {
```
Wow, this is next-level removing the animation for people who prefer reduced motion! Nice work!!! 🙌

meganmckissack replied:
Katia implemented this but I just learned about this as an option!

katia-utochkina replied:
Thank you😄. We thought it would be good to give users more control over our interface and detect when they prefer a reduced motion experience.

segdeha commented on src/App.css:
```diff
@@ -145,3 +160,91 @@ nav {
+  flex-direction: column;
+  align-items: center;
+  text-align: center;
+  max-width: 65rem;
```
`65rem` is...big. Is that to accommodate desktop?

katia-utochkina replied:
Yep, exactly that.

segdeha commented on src/App.css:
```diff
@@ -145,3 +160,91 @@ nav {
+  padding: 0;
+}
+
+.spinner {
```
I throttled my network and reloaded the app, but never saw the spinner 🤔 

Where is it meant to show up?

meganmckissack replied:
It should show up here the header is located, I tried it with 3G throttling and got it to show up.
![Screen Shot 2021-07-22 at 10 39 39 AM](https://user-images.githubusercontent.com/2355543/126685020-2e9770b8-10a1-4b77-8b63-1956e38ad535.png)

katia-utochkina replied:
@segdeha, you could see it briefly [here](https://www.loom.com/share/66cb33057add4d34a91aa8b3b5a72c77) 😄

segdeha replied:
I saw it! I had to throttle (which I had done) _and_ clear my cache. Looks great!

segdeha commented on src/App.css:
```diff
@@ -145,3 +160,91 @@ nav {
+  animation-duration: 1s;
+}
+
+@keyframes bounce {
```
Cool animation!

segdeha commented on src/App.js:
```diff
@@ -1,29 +1,32 @@
-    </Router>
+    <SearchResults>
+      <Router>
+        <div className="app">
```
Could we simplify this by removing the `<div>` and putting the class name on `<main>`?

katia-utochkina replied:
I don't think so -- especially with that nav bar bug we have,  the outer div needs to stay

segdeha commented on src/components/ErrorMessage.js:
```diff
@@ -0,0 +1,9 @@
+
+const ErrorMessage = ({ error }) => (
+  <div className="search__error">
+    <h2>{error ? 'Something went wrong' : 'Not found'}</h2>
```
I love the idea of surfacing an error when there’s, e.g., no network, but if there are simply no pins we could probably skip showing an error. What do you think?

katia-utochkina replied:
At the moment, we’re not showing errors if there’re no pins to render.
What gets displayed is a  map 
![no pins](https://user-images.githubusercontent.com/16920088/126694590-4381f5a7-3496-4d70-a0e3-e6b2292f00ce.png)

segdeha commented on src/components/Map.js:
```diff
@@ -18,10 +21,36 @@ const Map = () => {
             defaultZoom={zoom}
-          />
+            yesIWantToUseGoogleMapApiInternals={true}
+            onGoogleApiLoaded={handleApiLoaded}
```
👏👏👏

segdeha commented on src/components/Map.js:
```diff
@@ -18,10 +21,36 @@ const Map = () => {
+                      lat={coordinates[0].lat}
+                      lng={coordinates[0].lon}
+                      imageUrl={
+                        location?.thumbnail?.source ||
```
Again with the optional chaining!!! 💯

segdeha commented on src/components/Map.js:
```diff
@@ -18,10 +21,36 @@ const Map = () => {
+            yesIWantToUseGoogleMapApiInternals={true}
+            onGoogleApiLoaded={handleApiLoaded}
+          >
+            {data?.query?.pages?.length > 0 && loaded
```
Wow, love this use of [optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining)!!!

segdeha commented on src/components/Map.js:
```diff
@@ -18,10 +21,36 @@ const Map = () => {
+                      alt={
+                        location?.thumbnail?.source
+                          ? location.title
+                          : 'placeholder-image'
```
If there’s no image, I wonder if it would be a better experience for screen reader users just to leave the `alt` attribute blank?

katia-utochkina replied:
Good point, we thought about this, too. The ``<alt>`` tag  is blank now 😄

segdeha commented on src/components/Pin.js:
```diff
@@ -0,0 +1,13 @@
+const Pin = ({ imageUrl, alt }) => (
+  <div className="map-icon bounce">
+    <img src={imageUrl} alt={alt} />
+    <svg viewBox="0 0 32 32" aria-labelledby="caret-icon">
```
I love how simple this SVG is!

segdeha commented on src/context/SearchResults.js:
```diff
@@ -0,0 +1,35 @@
+    return <ErrorMessage error={error} />;
+  }
+
+  return isSuccess ? (
```
Great use of the context API, this component is so clean and clear!!!

segdeha commented on src/hooks/useFetch.js:
```diff
@@ -0,0 +1,37 @@
+      setStatus('loading');
+      try {
+        const res = await fetch(
+          `https://segdeha.com/api/nearby.php?lat=${lat}&lng=${lng}`,
```
Similar to @laurenmbeatty’s suggestion about keeping your `lat` and `lng` in just one place, could we do the same with this URL? What I’d suggest is having this be the only component that ever makes API calls, so line 13 in `SearchResults.js` could lose the 3rd condition (i.e. you only need a valid `lat` and `lng` to make the call. What do you think?

katia-utochkina replied:
I’ve removed the hardcoded url from the useFetch hook to solve the code duplication problem 😄

segdeha reviewed (changes requested):
This is fantastic work, y’all!!! I am so impressed, seriously! I did leave a few comments and suggestions that I would like to get your thoughts on before we merge, so I’m marking this as “request changes” (even if the resolution is just to answer the questions). Thanks!

katia-utochkina committed:
Remove hardcoded url from custom hook, make <alt> tag blank for placeholder images

alxmcr commented on src/hooks/useFetch.js:
```diff
@@ -0,0 +1,35 @@
+import { useState, useEffect } from 'react';
+
+const useFetch = (url, lat, lng) => {
```
Thank you so much for creating this custom hook! 😄

katia-utochkina replied:
@alxmcr , you asked for it 😄

laurenmbeatty commented on src/hooks/useFetch.js:
```diff
@@ -0,0 +1,35 @@
+    if (!url) return;
+
+    const getWikiEntries = async (lat, lng) => {
+      setStatus('loading');
```
I really like the use of this string literal for the status. 'loading' 'idle', and 'error'. Very human readable.

segdeha reviewed (approved):
Looks great! Such a thoughtful and thorough set of changes! Great work!!! 🎉

katia-utochkina committed:
Update file
